carma_guidance_plugins/include/carma_guidance_plugins/plugin_base_node.hpp

```diff
@@ -0,0 +1,189 @@
+    //! Publisher for plugin discovery. This publisher will begin publishing 
+    //immediately on node construction regardless of lifecycle state. This is meant to allow for fast plugin activation on startup.
+    rclcpp::Publisher<carma_planning_msgs::msg::Plugin>::SharedPtr plugin_discovery_pub_;
+
+    // Timers
+    //! Timer to trigger publication of the plugin discovery message at a fixed frequency 
+    rclcpp::TimerBase::SharedPtr discovert_timer_;
+
+    // WorldModel listener
+    carma_wm::WMListener wm_listener_;
+
+    // World Model populated by the listener at runtime
+    carma_wm::WorldModelConstPtr wm_;
+
+    /**
+     * \brief Callback for the plugin discovery timer which will publish the plugin discovery message
+     */
+    void discovery_timer_callback();
+
+  public:
+    /**
+     * \brief PluginBaseNode constructor 
+     */
+    explicit PluginBaseNode(const rclcpp::NodeOptions &);
+
+    //! Virtual destructor for safe deletion
+    virtual ~PluginBaseNode() = default;
+
+    /**
+     * \brief Returns the activation status of this plugin.
+     *        The plugins API callbacks will only be triggered when this method returns true. 
+     * 
+     * \return True if plugin is in the ACTIVE state. False otherwise.
+     */ 
+    virtual bool get_activation_status() final;
+
+
+    /**
+     * \brief Returns the type of this plugin according to the carma_planning_msgs::Plugin type enum.
+     *        Extending classes for the specific type should override this method. 
+     * 
+     * \return The extending class type or UNKOWN if the class or no parent class has implement this method. 
+     */ 
+    virtual uint8_t get_type();
+
+    /**
+     * \brief Get the availability status of this plugin based on the current operating environment.
+     *        Method must be overriden by extending classes. 
+     * 
+     * \return This method should return true if the plugin's current understanding of the world
+     *         means it would be capable of planning or executing its capability.
+     */ 
+    virtual bool get_availability() = 0;
+
+    /**
+     * \brief Get the capability string representing this plugins capabilities
+     *        Method must be overriden by extending classes.
+     *        Expectation is that abstract plugin type parent classes will provide a default implementation.
+     * 
+     * \return Forward slash separated string describing the plugin's capabilities per the plugin capabilites API
+     */ 
+    virtual std::string get_capability() = 0;
+
+    /**
+     * \brief Returns the name of this plugin.
+     *        This name may be automatically prepended to plugin API service or topic names. 
+     * 
+     * \return Name of this plugin
+     */ 
+    virtual std::string get_name() = 0;
+
+    /**
+     * \brief Returns the version id of this plugin.
+     * 
+     * \return The version id represented as a string
+     */ 
+    virtual std::string get_version_id() = 0;
```

Comment:
I can't recall, do we have specific format in mind for this version string? I know we use semver by convention, is it worth enforcing? I think we could do so with a non-virtual/final get_version_id() and pure virtual methods for get_major_version, get_minor_version, get_patch_version, or something like that.

Reply:
At the moment we don't have an enforced format anywhere. And as far as I know nothing actually uses this field. Probably best to wait until we have time to think about enforcing plugin versions.